### DISCUSSION TITLE
Magick::Image::Write and other symbols do link when using libc++ as the standard library

### DISCUSSION THREAD

**Discussion by FalcoGer** [ImageMagick/ImageMagick#6525] [Development] [answered]
### ImageMagick version

7.1.1-16 (Beta) Q16 x86_64 a0a5f3da4:20230730

### Operating system, version and so on

Ubuntu Mate 22.04

### Description

I have compiled ImageMagick like so

```text
CC=clang CXX=clang++ ./configure --enable-bounds-checking --disable-hdri --with-modules --with-utilities  --with-quantum-depth=16  --with-perl --with-x --with-dps --with-magick-plus-plus
make clean
make -j6
sudo make install
```

I have then written a program to test it. It works just fine, until I try to save the file to disk. I can create a Magick::Image, I can create Magick::Color objects, I can edit pixels, draw lines and text and many other things, and also display the image. But as soon as I try to write the image to disk, I get a linker error.

According to [this tutorial](https://imagemagick.org/Magick++/tutorial/Magick++_tutorial.pdf) you just do something like this

```
#include <Magick++.h>

int main()
{
    Magick::Image img{Magick::Geometry{480U, 300U}, "white"};
    img.write("white.png");
}
```

My IDE shows no warnings or errors, it compiles but doesn't link.

Compilation is done as such

```bash
sh -c 'clang++ -O2 -std=c++23 -stdlib=libc++ -L/usr/lib/llvm-17/lib `Magick++-config --libs --cppflags` test.cpp -o test'
```

Where `Magick++-config --libs --cppflags` evaluates to

```text
-fopenmp -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=16 -fopenmp -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=16 -fopenmp -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=16 -I/usr/local/include/ImageMagick-7
-L/usr/local/lib -lMagick++-7.Q16 -lMagickWand-7.Q16 -lMagickCore-7.Q16
```

The compiler output

```text
/usr/bin/ld: /tmp/mandelbrot-8a1ea0.o: in function `main':
mandelbrot.cpp:(.text+0x403): undefined reference to `Magick::Image::write(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

How do I link against libc++, because my actual use cases uses std::expected, which is not available in stdlibc++ yet?

**Comment by snibgo** [comment]
You use...
```
Magick++-config --libs --cppflags --cxxflags --ldflags
```
... but you show the output from ...
```
Magick++-config --libs --cppflags
```
Why don't you show the output of the complete command?

Have you tried this before building your program?
```
export PKG_CONFIG_PATH=${IM7DEV}../lib/pkgconfig
```
... where ${IM7DEV} is the installation directory for IM, and the `pkgconfig` directory contains a load of `*.pc` files. See https://imagemagick.org/script/magick++.php

**Reply by FalcoGer** [reply, marked as answer]
I solved the problem by recompiling ImageMagick against libc++

```bash
CC=clang CXX=clang++ CXXFLAGS="-stdlib=libc++ -L/usr/lib/llvm-17/lib -std=c++23 -O2" LDFLAGS="-stdlib=libc++ -L/usr/lib/llvm-17/lib" ./configure --enable-bounds-checking --disable-hdri --with-modules --with-utilities  --with-quantum-depth=16  --with-perl --with-x --with-dps --with-magick-plus-plus
make clean
make -j6
sudo make install
```

**Comment by FalcoGer** [comment]
> You use...
> 
> ```
> Magick++-config --libs --cppflags --cxxflags --ldflags
> ```
> 
> ... but you show the output from ...
> 
> ```
> Magick++-config --libs --cppflags
> ```
> 
> Why don't you show the output of the complete command?

Because I realized after examining the output of the command that --cppflags and --cxxflags are the same, and --ldflags and --libs are also the same. I forgot to update the compile command.
